chore(spark): update Spark examples README for batch jobs (#739)
kubeflow/sdk: examples/spark/README.md

This directory contains examples for using the Kubeflow Spark SDK.

> The Spark SDK supports two ways to run Spark, matching the [Spark SDK documentation](../../docs/source/spark/index.rst):

Reviewer: Nice writeup overall! One question — this and the other doc links in the file (batch-jobs.rst, lifecycle.rst, api.rst) point to the raw .rst source, which takes readers to unrendered text on GitHub rather than the built docs page. Since this README is aimed at users running examples, would it make sense to link to the rendered site instead, e.g. https://sdk.kubeflow.org/en/latest/spark/index.html? Curious if there's a reason to prefer .rst links here — happy to be wrong on this one

- **Interactive Sessions** - Connect to Spark from a notebook or script using Spark Connect.
- **Batch Jobs** - Submit existing Spark applications as managed Kubernetes workloads.

For the full batch job documentation, see [Batch Jobs](../../docs/source/spark/batch-jobs.rst) and [Job Lifecycle](../../docs/source/spark/lifecycle.rst).

## Examples

### Interactive Sessions

- **spark_connect_simple.py** - Basic SparkClient usage with simple API
- **spark_advanced_options.py** - Advanced configuration with Driver/Executor objects
- **demo_existing_sparkconnect.py** - Connect to existing SparkConnect cluster
- **connect_existing_session.py** - Connect to an existing Spark Connect session through `base_url`
- **test_connect_url.py** - Test URL-based connection to Spark Connect

### Batch Jobs

- **batch_job_lifecycle.py** - Submit a `FileJob` and exercise the batch job lifecycle (`get_job`, `list_jobs`, `get_job_logs`, `delete_job`)
- **batch_func_job_lifecycle.py** - Submit a `FuncJob` and exercise the batch job lifecycle
- **batch_failed_job.py** - Submit a `FileJob` expected to fail and inspect the `FAILED` state

> - **batch_job_options.py** - Submit a batch job with Kubernetes options (labels, annotations, node selector, tolerations, custom name)

Reviewer: Nit: since batch_job_options.py demonstrates Kubernetes options (labels, annotations, node selector, tolerations), might be worth also linking the Options Reference doc here or in the Batch Job APIs section below — right now the README links batch-jobs.rst, lifecycle.rst, and api.rst, but not options.rst, even though it's the page that documents exactly what this example shows.

This directory contains examples for using the Kubeflow Spark SDK.

The Spark SDK supports two ways to run Spark, matching the [Spark SDK documentation](https://sdk.kubeflow.org/en/latest/spark/index.html):

- **Interactive Sessions** - Connect to Spark from a notebook or script using Spark Connect.

> - **Batch Jobs** - Submit existing Spark applications as managed Kubernetes workloads.

Reviewer 2: ```suggestion
- **Interactive Sessions** - Connect to Spark from a notebook or script using Spark Connect.
- **Batch Jobs** - Submit Spark applications as managed Kubernetes workloads.
```

For the full batch job documentation, see [Batch Jobs](https://sdk.kubeflow.org/en/latest/spark/batch-jobs.html) and [Job Lifecycle](https://sdk.kubeflow.org/en/latest/spark/lifecycle.html).

> 

Reviewer 2: If we add the batch job documentation it would be nice to add the doc [link](https://sdk.kubeflow.org/en/latest/spark/sessions.html) for interactive session as well.

### Batch Jobs

- **batch_job_lifecycle.py** - Submit a `FileJob` and exercise the batch job lifecycle (`get_job`, `list_jobs`, `get_job_logs`, `delete_job`)
- **batch_func_job_lifecycle.py** - Submit a `FuncJob` and exercise the batch job lifecycle
- **batch_failed_job.py** - Submit a `FileJob` expected to fail and inspect the `FAILED` state
- **batch_job_options.py** - Submit a batch job with Kubernetes options (labels, annotations, node selector, tolerations, custom name)
  See the [Options Reference](https://sdk.kubeflow.org/en/latest/spark/options.html) for details on Kubernetes options.
- **spark_job.py** - A simple Spark application used as the remote `file_source` for the batch job examples

## Prerequisites

Install spark dependencies:
```bash
uv pip install kubeflow[spark]
```

Batch job examples run against a Kubernetes cluster with the Spark Operator installed. Batch job submission requires a `spark-operator-spark` ServiceAccount in the target namespace with the required SparkApplication RBAC permissions. See the [Spark SDK docs](https://sdk.kubeflow.org/en/latest/spark/index.html) for prerequisites.

> 

Reviewer 2: ```suggestion
The Spark examples run against a Kubernetes cluster with the Spark Operator installed. Batch job submission requires a `spark-operator-spark` ServiceAccount in the target namespace with the required SparkApplication RBAC permissions. See the [Spark SDK docs](https://sdk.kubeflow.org/en/latest/spark/index.html) for prerequisites.

```

## Running Examples

> 
> ```bash
> # Run from repository root
> uv run python examples/spark/spark_connect_simple.py

Reviewer 2: ```suggestion
# Run from repository root

# Run an interactive session example
uv run python examples/spark/spark_connect_simple.py
```

# Run a batch job lifecycle example
uv run python examples/spark/batch_job_lifecycle.py
```

> 

Reviewer 2: Since we add the batch job APIs below it might be good to have a section here to add interactive session APIs as well.

  See the [Options Reference](https://sdk.kubeflow.org/en/latest/spark/options.html) for details on Kubernetes options.
- **spark_job.py** - A simple Spark application used as the remote `file_source` for the batch job examples

## Prerequisites

Install spark dependencies:
```bash
uv pip install kubeflow[spark]
```

The Spark examples run against a Kubernetes cluster with the Spark Operator installed. Batch job submission requires a `spark-operator-spark` ServiceAccount in the target namespace with the required SparkApplication RBAC permissions. See the [Spark SDK docs](https://sdk.kubeflow.org/en/latest/spark/index.html) for prerequisites.

## Running Examples

```bash
# Run from repository root

# Run an interactive session example
uv run python examples/spark/spark_connect_simple.py

# Run a batch job lifecycle example
uv run python examples/spark/batch_job_lifecycle.py
```

## Interactive Session APIs

Interactive session examples use `SparkClient` and `connect()`:

- `SparkClient()` - Create a client for the Kubernetes cluster
- `connect()` - Create a new Spark Connect session, or connect to an existing server when `base_url` is provided

> - `delete_session()` - Delete a Spark Connect session when you are done with it

Reviewer 2: Just to ensure consistency with how we portray the batch job APIs below.

```suggestion
- `connect()` - Create a new Spark Connect session, or connect to an existing server when `base_url` is provided
- `list_sessions()` / `get_session()` - Inspect Spark Connect sessions
- `get_session_logs()` - Read session logs
- `delete_session()` - Delete a Spark Connect session when you are done with it
```

See the [Spark SDK API reference](https://sdk.kubeflow.org/en/latest/spark/api.html) for the current API surface.

## Batch Job APIs

Batch job examples use `submit_job()` with either `FileJob` or `FuncJob`, then manage the job with the lifecycle APIs:

> - `submit_job()` - Submit a Spark application as a managed Kubernetes workload
> - `get_job()` / `list_jobs()` - Inspect batch jobs
> - `wait_for_job_status()` - Wait for a job to reach a target `SparkJobStatus`
> - `get_job_logs()` - Stream driver pod logs
> - `delete_job()` - Delete a batch job

Reviewer 2: Streaming the logs is an optional capability.

```suggestion
- `submit_job()` - Submit a Spark application as a managed Kubernetes workload
- `get_job()` / `list_jobs()` - Inspect batch jobs
- `wait_for_job_status()` - Wait for a job to reach a target `SparkJobStatus`
- `get_job_logs()` - Read driver pod logs
- `delete_job()` - Delete a batch job
```
